src/lib/operators/insert.cpp

```diff
@@ -282,25 +284,13 @@ void Insert::_on_rollback_records() {
 
     for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
          ++chunk_offset) {
-      mvcc_data->set_end_cid(chunk_offset, CommitID{0});
-
-      // Update chunk statistics.
-      target_chunk->increase_invalid_row_count(ChunkOffset{1});
-    }
-
-    // This fence guarantees that no other thread will ever observe `begin_cid = 0 && end_cid != 0` for rolled-back
-    // records.
-    std::atomic_thread_fence(std::memory_order_release);
-
-    for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
-         ++chunk_offset) {
-      mvcc_data->set_begin_cid(chunk_offset, CommitID{0});
+      mvcc_data->set_end_cid(chunk_offset, CommitID{0}, std::memory_order_seq_cst);
+      mvcc_data->set_begin_cid(chunk_offset, CommitID{0}, std::memory_order_seq_cst);
```

Comment:
Are you 100% sure that this means the end CIDs are really set before the begin CIDs (and TIDs) and that the memory order does not only guarantee writes to the single atomics? I think that's what the two loops with fence in between tried to ensure

Reply:
I will run benchmarks, but as long as there is no regression I would keep it this way to be safe. I don't think I'll have to the time to fully dive into MVCC and atomics. :-/

Reply:
I would also opt for safety if performance is not getting too bad